pkg/lightning/backend/local.go

```diff
@@ -1830,34 +1882,38 @@ outside:
 		select {
 		case kvs, ok := <-w.kvsChan:
 			if !ok {
+				w.markCleanQueue()
 				break outside
 			}
 
 			w.writeBatch.append(kvs)
-			if w.writeBatch.totalSize <= w.memtableSizeLimit {
-				break
-			}
-			if w.writer == nil {
-				w.writer, err = newSSTWriter(w.genSSTPath())
-				if err != nil {
+			w.markDirtyPost()
+
+			if w.writeBatch.totalSize > w.memtableSizeLimit {
+				if w.writer == nil {
+					w.writer, err = newSSTWriter(w.genSSTPath())
+					if err != nil {
+						w.writeErr.Set(err)
+						return
+					}
+				}
+
+				if err = w.writeKVsOrIngest(0); err != nil {
```

Comment:
What if we do `w.markCleanFlush()` after this ingest and the ingest  below. If after this ingest, there is no data in the channel, this should be a consistent state that can do save the checkpoint.

Reply:
> this should be a consistent state that can do save the checkpoint.

that's impossible. `writeKVsOrIngest` does one of the two things:

1. move the content of `w.writeBatch` into `w.writer` → so `w.writeBatch` will become empty but `w.writer` has content, or
2. directly ingest `w.writeBatch` into Pebble because it is not sorted after `w.writer` → a necessary condition for this is `w.writer` not being empty, thus the LocalWriter is still dirty.

Reply:
What about `writeKVsOrIngest ` return a bool to represent the flush state, if it triggers a ingest, then we can do a `markCleanFlush`.

Reply:
@glorv both cases 1 and 2 are dirty.

Reply:
I think there is no conflict if we do `w.dirtyState.Add(-2 * flushedKvCount)`. 
In case 1, move content into w.writer, this won't change dirtyState, so It's still dirty.
In case 2, the decrease won't contain the kvs that still in w.writer so it's still dirty too.

A clean state can be defined as `All kvs send by AppendRows are by ingested`, so if we only decrease the count after ingest by the ingested kv count, we can make sure if the count decrease to 0, the state must be clean.